napari/_vispy/vispy_base_layer.py

```diff
@@ -192,10 +190,108 @@ def _reset_base(self):
+        threshold.
+
+        Parameters
+        ----------
+        size : tuple
+            Requested size of field of view in data coordinates
+        size_threshold : int, optional
+            Maximum size of a displayed tile in pixels. If not provided then
+            taken from the size of the canvas.
+
+        Returns
+        ----------
+        level : int
+            Level of the pyramid to be viewing.
+        """
+
+        # Determine size threshold
+        if size_threshold is None:
+            size_threshold = np.max(self.node.canvas.size)
+
+        # Only consider displayed dimensions
+        size = size[self.layer.dims.displayed]
+        ds = self.layer.level_downsamples[:, self.layer.dims.displayed]
+
+        # Consider largest dimension after scaling by downsamples
+        ratio = (size / ds).max(axis=1)
+
+        # Find the maximum allowed size
+        level = np.argmin(ratio > size_threshold)
+
+        return level
+
+    def find_coordinates_of_corners_of_canvas(self):
+        """Find location of the corners of canvas in data coordinates.
+
+        Depends on the current pan and zoom position.
+
+        Returns
+        ----------
+        corner_pixels : array
+            Coordinates of top left and bottom right canvas pixel in the data.
+        size : array
+            Size of requested tile in data coordinates.
+        """
+        nd = self.layer.dims.ndisplay
+        # Find image coordinate of top left canvas pixel
+        if self.node.canvas is not None:
+            offset = self.translate[:nd] / self.scale[:nd]
+            tl_raw = np.floor(self._transform_position([0, 0]) + offset[::-1])
+            br_raw = np.ceil(
+                self._transform_position(self.node.canvas.size) + offset[::-1]
+            )
+        else:
+            tl_raw = [0] * nd
+            br_raw = [1] * nd
+
+        top_left = np.zeros(self.layer.ndim)
+        bottom_right = np.zeros(self.layer.ndim)
+        for d, tl, br in zip(self.layer.dims.displayed, tl_raw, br_raw):
+            top_left[d] = tl
+            bottom_right[d] = br
+
+        # Clip according to the max data of the level shape
+        top_left = np.clip(
+            top_left,
+            0,
+            np.subtract(self.layer.level_shapes[self.layer.data_level], 1),
+        )
+        bottom_right = np.clip(
+            bottom_right,
+            0,
+            np.subtract(self.layer.level_shapes[self.layer.data_level], 1),
+        )
+
+        # Scale to full resolution of the data
+        size = (bottom_right - top_left) * self.layer.level_downsamples[
+            self.layer.data_level
+        ]
+
+        return np.array([top_left, bottom_right]).astype(int), size
+
     def on_draw(self, event):
-        """Called whenever the canvas is drawn.
+        """Called whenever the canvas is drawn, which happens whenever new
+        data is sent to the canvas or the camera is moved.
```

Comment:
```suggestion
        """Called whenever the canvas is drawn.
        
         This is triggered whenever new data is sent to the canvas
         or the camera is moved.
```

how is this triggered? If it's an automagic thing from VisPy, it's worth pointing it out. If it's some place in our code, that is also worth pointing out.

Reply:
done